posit-dev/positron
# Add a Positron API for loading URLs in the Viewer pane

### Intent

This change adds an API to Positron for previewing a URL. It also adds a mini web browser to the Viewer pane, which can be used to load various kinds of interactive web content. 

<img width="1282" alt="image" src="https://github.com/posit-dev/positron/assets/470418/03aa89fc-f829-43b9-9f1e-66e23f226661">

Addresses https://github.com/posit-dev/positron/issues/814

Also addresses:
- https://github.com/posit-dev/positron/issues/1304
- https://github.com/posit-dev/positron/issues/2090 (via https://github.com/posit-dev/amalthea/pull/273)

### Approach

The new web browser uses Positron's existing UI components, and is somewhat loosely based on Simple Browser. The major parts of this change are:

- a new API for previewing a URL; it accepts a URI and returns a `PreviewPanel` so extensions that create previews can follow up by e.g. sending messages to the preview window
- a new built-in extension, `positron-viewer`; it registers itself as a URI opener and uses the new API to offer to load URLs in the Viewer pane where appropriate
- new rules for detecting links in the Console so that we can offer to open them in the Viewer pane
- a new web browser toolbar that uses native Positron UX components and window messaging to navigate/control the Viewer pane
- a new UI comm message that can be used to open content in the Viewer pane directly; this is how Shiny works in the example above

### QA Notes

It is worth noting that this new URL opener is registered globally; you can even open content from the Terminal in the Viewer pane by using Cmd+Click on URLs there. 

Positron shouldn't offer to open content in the Viewer pane unless it is local (i.e. has a URL that looks like it refers to a localhost interface). 

You can't just mash any URL you want into the address bar; external URLs will not work. We'll add better behavior here, but for now if you enter a syntactically invalid URL (which I know you are going to try), you'll get an error, and if you enter a well-formed URL we can't load, you will just get a blank viewer.

This change adds a new URL opener and there's been some debate concerning how aggressive that opener should be. In this PR I've taken the most conservative approach, which is to offer users a choice every time. In the future we will want to either provide a more opinionated default or provide better UI than what VS Code does for setting the default.
